src/rust/bitbox02/src/testing.rs

```diff
@@ -0,0 +1,34 @@
+// Copyright 2020 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Small mocking infrastructure for testing.
+
+extern crate std;
+use std::sync::{Arc, Mutex};
+
+#[derive(Default)]
+pub struct Data {
+    pub memory_set_device_name_arg: Option<std::string::String>,
+    pub memory_set_device_name_result: Option<Result<(), ()>>,
+    pub ui_confirm_create_body: Option<std::string::String>,
+    pub ui_confirm_create_result: Option<bool>,
+}
+
+lazy_static! {
+    pub static ref TESTDATA: Arc<Mutex<Data>> = Arc::new(Mutex::new(Default::default()));
+}
+
+pub fn mock(data: Data) {
+    *TESTDATA.lock().unwrap() = data;
```

Comment:
I'm not entirely sure putting this behind a mutex is a good idea. I mean, by default rust tests run concurrencly, right. So, this makes sense. But now imagine 2 tests where each one updates TESTDATA. The end result is we might have failing tests.

Am I missing something?

Reply:
No you are right I think. So I need to keep the lock for the duration of the test to avoid a race condition, right? Make it a RW-lock too, as the stub itself also accesses the testdata.

Reply:
Scratch that, makes no sense, as the read lock cannot be acquired while it is write locked...

Reply:
So, generally speaking we need to serialize all tests which access `TESTDATA`. There are different ways to do it but I haven't come up with what's best yet.

Alternatively, we could run each test which access `TESTDATA` as a separate binary but it feels like overkill.

Reply:
And yet another way is to have separate `TESTDATA` for each test which needs it, like `TESTDATA_set_device_name`.

Reply:
Exactly...

Reply:
added a new iteartion, PTAL:

1) global mutex to serialize tests
2) must be held when accessing DATA, which is now in a RefCell instead of a Mutex (runtime borrow checking, but no thread safety).